lib/has_array_of/associated_array.rb

```diff
@@ -26,19 +26,19 @@ def self.define_in(owner_model, options)
               else
                 [first, *rest]
               end
-        "ARRAY[#{ary.map(&try_pkey_proc).join(',')}]::#{pkey_attribute_sql_type}[]"
+        ary.map(&try_pkey_proc)
       end
 
       define_singleton_method "with_#{name}_containing" do |*args|
-        where "#{ids_attribute} @> #{expression[args]}"
+        where "#{ids_attribute} @> ARRAY[?]::#{pkey_attribute_sql_type}[]", expression[args]
```

Comment:
Please extract `"#{ids_attribute} @> ARRAY[?]::#{pkey_attribute_sql_type}[]"` string to a `with_containing_condition` variable and use it in a closure. Doing so we'll avoid a string allocation :)

Reply:
Do you want to extract only `ARRAY[?]::#{pkey_attribute_sql_type}[]"` or do smth like `"#{ids_attribute} #{expression} ARRAY[?]::#{pkey_attribute_sql_type}[]"`?

Reply:
I don't really care how this would be actually achieved but the key thing is to not allocate a string inside the block. Something like this:

```ruby
with_containing_condition = "#{ids_attribute} @> ARRAY[?]::#{pkey_attribute_sql_type}[]".freeze
define_singleton_method "with_#{name}_containing" do |*args|
  where(with_containing_condition, expression[args])
end

with_contained_in_condition = "#{ids_attribute} <@ ARRAY[?]::#{pkey_attribute_sql_type}[]".freeze
define_singleton_method ...

with_any_from_condition = "#{ids_attribute} && ARRAY[?]::#{pkey_attribute_sql_type}
 []".freeze
define_singleton_method ...
```

Freezing probably is not mandatory but it clearly shows the intention to define a "constant".
In fact, "#{ids_attribute} @> ARRAY[?]::#{pkey_attribute_sql_type}[]" is not being changed in runtime so we can calculate it just once and leave out of method body.

Do you abstract the `#{expression}` or not — i don't really care, it's a matter of taste. IMO, it's not worth that.